packages/core/src/utils/operation.ts

```diff
@@ -0,0 +1,44 @@
+import {
+  GraphQLRequest,
+  Operation,
+  OperationContext,
+  OperationType,
+} from '../types';
+import { Warning, deprecationWarning } from './deprecation';
+
+// TODO: Remove when the deprecated `operationName` property is removed
+const DEPRECATED: Record<string, Warning> = {
+  operationName: {
+    key: 'Operation.operationName',
+    message:
+      'The "Operation.operationName" property has been deprecated and will be removed in a future release of urql. Use "Operation.kind" instead.',
+  },
+};
+
+export const makeOperation = (
+  kind: OperationType,
+  {
+    key,
+    query,
+    variables,
+  }: Pick<GraphQLRequest, 'key' | 'query' | 'variables'>,
+  context: OperationContext
+): Operation => ({
+  key,
+  query,
+  variables,
+  kind,
+  context,
+
+  get operationName(): OperationType {
+    deprecationWarning(DEPRECATED.operationName);
+
+    return this.kind;
+  },
+
+  set operationName(kind: OperationType) {
+    deprecationWarning(DEPRECATED.operationName);
+
+    this.kind = kind;
+  },
```

Comment:
I'd just drop this, as setting isn't ever expected; we have a couple of edge cases of course that we went through (e.g. spreading) but even then `operationName` is pretty much readonly. I think this also hints at the fact that we should update the `Operation` type to make `kind` competely `readonly`